### ISSUE TITLE
CRI: Handle streaming requests (attach/exec/port-forward)

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#29579]
I re-titled the issue since the discussion has naturally shifted to the general design and implementation of streaming requests. 
**Please read the summary doc before joining the discussion. Thanks!**
https://docs.google.com/document/d/1MreuHzNvkBW6q7o_zehm1CBOBof3shbtMTGtUpjpRmY/edit?usp=sharing

------ The original post -----
#25273 added terminal resizing for run/attach/exec. We need to support this with the new container runtime interface.

Current implementations of terminal resizing:
docker (both `nsenter` and `docker exec`): https://github.com/kubernetes/kubernetes/blob/3b21a9901bcd48bb452d3bf1a0cddc90dae142c4/pkg/kubelet/dockertools/exec.go
rkt: https://github.com/kubernetes/kubernetes/blob/3b21a9901bcd48bb452d3bf1a0cddc90dae142c4/pkg/kubelet/rkt/rkt.go#L2018

The new container interface/api (https://github.com/kubernetes/kubernetes/blob/5faa9b1283c807e9af07a7cc119d5c1a30fec31c/pkg/kubelet/api/v1alpha1/runtime/api.proto#L570) has only a `Exec` method with no resizing support.

@ncdc, any suggestion of how we could achieve this? I think it may make sense to add this to the interface and push it down to the runtime. 

/cc @kubernetes/sig-node

**Comment by ncdc** [comment]
It has to be pushed down to the runtime given that the docker exec
implementation is different from rkt.

**Comment by euank** [comment]
The only other option I can think of is the interface being a pty, in which case the resizes would not be pushed down to the runtime, though initial pty creation would be.

**Comment by yujuhong** [comment]
> The only other option I can think of is the interface being a pty-path, in which case the resizes would not be pushed down to the runtime, though initial pty creation would be.

Hm... I don't think this'd work with hyper. /cc @feiskyer 

The implementation that uses `docker exec` today does:
 * call docker CreateExec
 * start the `HandleResizing` goroutine with docker `ResizeExecTTY`
 * call docker `StartExec`.

That means to support terminal resizing in the runtime interface, we'll need to replace the current `Exec` with `CreateExec` and `StartExec`, while also adding `ResizeExecTTY`. Is that correct? @ncdc @euank 
I am not too keen on the change, but I also don't see other alternatives. Any suggestions?

**Comment by feiskyer** [comment]
Hyper supports same APIs like docker: `CreateExec`, `ResizeTTY` and `StartExec`. The problem is could we suppose all runtimes support these three apis? Is it work for rkt and oci-runtime?

> The only other option I can think of is the interface being a pty-path, in which case the resizes would not be pushed down to the runtime, though initial pty creation would be.

Could we just add a new terminal resize api to new runtime API, e.g. `ResizeTTY`, and let runtime itself decides how to implements the tty resizing?

**Comment by ncdc** [comment]
I'm traveling today - please ping me on Monday so I don't forget to reply.
Thanks!

**Comment by yujuhong** [comment]
> I'm traveling today - please ping me on Monday so I don't forget to reply.
> Thanks!

@ncdc, to help with tracking, let's assign this to you until you reply.

**Comment by yujuhong** [comment]
As instructed, @ncdc, here is a friendly ping to remind you to update the issue!

**Comment by euank** [comment]
The interface I was trying to suggest is an alternative is the following:

1) kubelet calls `Exec(interactive=true,initialWidth=x,initialHeight=y)` or maybe `ExecInteractive` or whatever
1a) runtime creates a pty and connects the slave end to the runtime somehow (whether by controlling it and proxying through to a `CreateExec`, `StartExec`, and `HandleResizing` or by just connecting to an existing pty or constructing one from some pipes it has laying around)
1b) runtime returns the path to the pty on the host
2) kubelet opens the pty to read, write, and resize
3) kubelet closes pty

Any runtime that can actually support attach in any manner should be able to construct a pty, even if it has to act as the slave side of it and translate it back into runtime api calls on the other end.

That feels like a slightly cleaner interface than `create/resize/exec`.

**Comment by ncdc** [comment]
@euank how would a call in the kubelet to resize the pty translate to a remote call to Docker for `ResizeExecTTY`?

**Comment by yujuhong** [comment]
> @euank how would a call in the kubelet to resize the pty translate to a remote call to Docker for ResizeExecTTY?

+1. I am not sure how this would work.

**Comment by euank** [comment]
@ncdc in this case the "runtime" would be the `docker-shim` package within the kubelet process, which would hold the slave end of the pty and then execute the ResizeExecTTY call against docker.

**Comment by ncdc** [comment]
@euank ok, trying to reason through this... so `kubectl` receives `SIGWINCH` and sends a message to the Kubelet that the container's terminal needs to be resized to 100x50. Does the Kubelet then call `term.SetSize(ptyFD, 100x50)`? How can we, in Go, react to the `term.SetSize` call and a) for Docker, call `ResizeExecTTY`, b) for rkt, do whatever is appropriate for rkt, etc?

cc @mrunalp

**Comment by yujuhong** [comment]
@ncdc @euank @feiskyer can we just add a resize stream in attach/exec in CRI, and let kubelet proxy the resize requests to the runtime? That way, runtime can handle the resize stream however it likes. Does this make sense?

**Comment by feiskyer** [comment]
> can we just add a resize stream in attach/exec in CRI, and let kubelet proxy the resize requests to the runtime? That way, runtime can handle the resize stream however it likes. Does this make sense?

+1. (Same idea as https://github.com/kubernetes/kubernetes/issues/29579#issuecomment-236072601.)

**Comment by ncdc** [comment]
I need to look at the CRI in more detail when I'm at my desk but that
sounds reasonable.

**Comment by yujuhong** [comment]
> +1. (Same idea as #29579 (comment).)

What I suggested is a little different than your comment, which was to add a `ResizeTTY` call :)
I am suggesting that we add a new "resize" stream in the `Exec` call (in additional to stdin stream), so that kubelet can just forward the resize request directly. That way, we don't have to split `Exec` to `CreateExec` and `StartExec` to get the exec ID that can be used to call `ResizeTTY`. Of course, the runtime will need to spawn a thread reading the resize stream and react (via calling `ResizeTTY` or something else) to it.

**Comment by yujuhong** [comment]
> What I suggested is a little different than your comment, which was to add a ResizeTTY call :)
> I am suggesting that we add a new "resize" stream in the Exec call (in additional to stdin stream), so that kubelet can just forward the resize request directly. That way, we don't have to split Exec to CreateExec and StartExec to get the exec ID that can be used to call ResizeTTY. Of course, the runtime will need to spawn a thread reading the resize stream and react (via calling ResizeTTY or something else) to it.

It could be as simple as https://github.com/yujuhong/kubernetes/commit/355a9cec74f62fba3694d4eebff8553b8ef9a2ca for CRI. The implementation is another story.

**Comment by feiskyer** [comment]
@yujuhong I see. This could keep the CRI simple.

**Comment by yujuhong** [comment]
@ncdc @euank any thoughts on https://github.com/yujuhong/kubernetes/commit/355a9cec74f62fba3694d4eebff8553b8ef9a2ca?

**Comment by ncdc** [comment]
@yujuhong that seems reasonable. I don't have much familiarity with protobuf - will we be able to treat `Stdin` and `Resize` as `io.Reader`s?

**Comment by yujuhong** [comment]
As discussed in the sig-node meeting today, how stdin/stdout/stderr streams are handled in the proto file have some issues. We need to fix that.

/cc @smarterclayton @feiskyer

**Comment by smarterclayton** [comment]
I commented on the old PR, but we know for these we'll need a multiplexed multi-channel stream with non-blocking.  That's http2 - I'm ok saying that's our new model, but we'll have to deal with the fact that we want to have the API server be able to connect to the directly (kubelet can't be a network bandwidth hog).  So exec flow is really different than all other calls - ideally it's:

1. apiserver talks to kubelet
2. Kubelet ensures that CRI knows an exec session is incoming, and returns a URL that the kubelet can pass back to the APIserver
3. APIServer direct connects to the info the kubelet passed (which points to the exec stub in container)
4. APIServer talks HTTP2 to the exec stub.

**Comment by yujuhong** [comment]
@smarterclayton I don't have prior experience with protobuf/grpc, so here what I've read so far. grpc is built on http2, but it streams messages (not arbitrary streams).
We can emulate a stream by sending out a chunk of data encapsulated in a message.
E.g., 
```
message DataChunk {
    bytes data = 1;
}
```

I think kubelet should also be able to read messages, and demultiplex the fields in the message and send them back apiserver via different streams. There will definitely incur overhead on the kubelet's side but it should be doable(?). What do you think about it?
```
message ExecResponse {
    bytes stdout = 1;
    bytes stderr = 2;
}
```

**Comment by smarterclayton** [comment]
We are up to five exec streams, and I suspect more are needed.  I
think grpc may not be the right solution for attach, exec, and
portforward, and we may prefer to continue down our current path where
we have good go support for this problem already.  Ultimately API
server has to support this protocol, and clients to.  I don't want to
invent a brand new protocol now - I think it's much more reasonable to
evolve our existing protocol.  There is at least 6 dev months invested
in our current approach, so we shouldn't underestimate the challenge.

I'm cautious about logs being possible under grpc, but I think this
topic deserves a serious design proposal that focuses on the right
system behavior before we decide grpc is it.

**Comment by yujuhong** [comment]
> I
> think grpc may not be the right solution for attach, exec, and
> portforward, and we may prefer to continue down our current path where
> we have good go support for this problem already.

We are not committed to grpc if exec, attach, and portforward can't be supported easily. 

> I'm cautious about logs being possible under grpc, but I think this
> topic deserves a serious design proposal that focuses on the right
> system behavior before we decide grpc is it.

Are you talking about getting container logs? The current rkt integration does this already. Whether it's efficient or not is a separate question.


There are several options. I believe we haven't invested so much in grpc that we can't swap it out completely. There is also a possibility of following docker's (and containerd's) approach to [add paths to stdin/stdout/stderr in the request](https://github.com/docker/containerd/blob/9df37c23e4cdebcd3c37bc5448b583302133cb15/api/grpc/types/api.proto#L78). The implementation of Exec will be tied closely to how we are going to support attach and portforward too. 

Let's find a time to discuss this over a meeting. 

/cc @thockin @dchen1107

**Comment by feiskyer** [comment]
@smarterclayton `Exec` is already a stream, see https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/runtime/api.proto#L42, and other interfaces could also be added as grpc stream. 

A basic reference of grpc stream could be found at http://www.grpc.io/docs/tutorials/basic/go.html#server-side-streaming-rpc.

**Comment by yujuhong** [comment]
> I think kubelet should also be able to read messages, and demultiplex the fields in the message and send them back to apiserver via different streams.

@smarterclayton I'm trying to understand whether the overhead of this is significantly different than handling a typical http stream (i.e., copying between buffers). What are your thoughts?

**Comment by feiskyer** [comment]
> We are up to five exec streams, and I suspect more are needed.  I
> think grpc may not be the right solution for attach, exec, and
> portforward, and we may prefer to continue down our current path where
> we have good go support for this problem already.  Ultimately API
> server has to support this protocol, and clients to.  I don't want to
> invent a brand new protocol now - I think it's much more reasonable to
> evolve our existing protocol.  There is at least 6 dev months invested
> in our current approach, so we shouldn't underestimate the challenge.

@smarterclayton What's the problem of defining separate grpc streaming APIs for attach, exec,..., e.g.

```
rpc Exec(stream ExecRequest) returns (stream ExecResponse) {}
rpc Attach(stream AttachRequest) returns (stream AttachResponse) {}
rpc Logs(LogsRequest) returns (stream LogsResponse) {}
```

btw, if you are saying multi-channel stream (e.g. separate stdout and stderr), it could be achieved by (while ExecResponse is a stream response):

```
message ExecResponse {
    bytes stdout = 1;
    bytes stderr = 2;
}
```

> I'm cautious about logs being possible under grpc, but I think this
> topic deserves a serious design proposal that focuses on the right
> system behavior before we decide grpc is it.

Logs interface only streams output, so its impl looks like `ListFeatures` at http://www.grpc.io/docs/tutorials/basic/go.html#server-side-streaming-rpc.

**Comment by smarterclayton** [comment]
Exec has to accept multiple streams, as well as returning multiple streams,
without HoL blocking ideally.

**Comment by smarterclayton** [comment]
It does look like gRPC support bidirectional streams, but focusing just on
gRPC is only half the picture.  The client needs a protocol to connect to
the master, which needs a protocol to connect to the kubelet, which to be
correct, needs to redirect the master connection directly to a process
inside of the container.

The protocol used by the client and master, or master and kubelet today is
our SPDY or WS protocol.  We do not want to continue sending attach and
exec through the kubelet, nor do we want the runtime daemons proxying this
data in the long term (that means the network and CPU bandwidth defined by
exec are not accounted to the container).  The kubelet needs to authorize
the request and then hand it off to the container.  So ultimately, the
streaming protocol can be completely separated from the transaction of
establishing the transaction protocol.

I don't think CRI should define exec/attach/portforward *streaming* - it
needs to define establishing that streaming, and then let the container
shim process manage the subsequent connection.

**Comment by mrunalp** [comment]
@smarterclayton That implies that we want to follow the docker model and use pipes. The API could then return the path to the pipes as @yujuhong suggested earlier.

**Comment by vishh** [comment]
How will kubelet inject a binary that can handle exec and attach and invoke that binary with runtime's help? And doesn't attach require holding the pty fd?

**Comment by yujuhong** [comment]
> I don't think CRI should define exec/attach/portforward *streaming* - it
> needs to define establishing that streaming, and then let the container
> shim process manage the subsequent connection.

+1 for that.

> How will kubelet inject a binary that can handle exec and attach and invoke that binary with runtime's help? And doesn't attach require holding the pty fd?

Discussed with @vishh offline yesterday.

If we can inject a binary into the container to handle streaming operations (interactive exec, attach, and portforward), kubelet can just `exec` that binary and then redirect the master connection to the process. This might be more portable across runtimes (esp. for hypervised runtimes), and CRI wouldn't have to support any additional calls except `exec`. I think this is aligned with what @smarterclayton suggested earlier for port-forwarding https://github.com/kubernetes/kubernetes/issues/25113#issuecomment-216717692

**Comment by smarterclayton** [comment]
Yes, although details may be deeper, I think the right approach is to get a
static / custom binary into the container that allows execution and is
container runtime agnostic. Ultimately that means we can deliver other
useful debugging tools that work cross container runtime (list filesystem,
fetch files, get core dumps) over time.

On Aug 10, 2016, at 5:11 PM, Yu-Ju Hong <notifications@github.com> wrote:

I don't think CRI should define exec/attach/portforward *streaming* - it
needs to define establishing that streaming, and then let the container
shim process manage the subsequent connection.

+1 for that.

How will kubelet inject a binary that can handle exec and attach and invoke
that binary with runtime's help? And doesn't attach require holding the pty
fd?

Discussed with @vishh <https://github.com/vishh> offline yesterday.

If we can inject a binary into the container to handle streaming operations
(interactive exec, attach, and portforward), kubelet can just exec that
binary and then redirect the master connection to the process. This might
be more portable across runtimes (esp. for hypervised runtimes), and CRI
wouldn't have to support any additional calls except exec. I think this is
aligned with what @smarterclayton <https://github.com/smarterclayton>
suggested earlier for port-forwarding #25113 (comment)
<https://github.com/kubernetes/kubernetes/issues/25113#issuecomment-216717692>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/29579#issuecomment-239005819>,
or mute the thread

**Comment by euank** [comment]
@smarterclayton I'm not sure I get how your proposal works with attach. Is it explicitly suggesting that the init process of each container is overridden by a new process which understands some kubelet-defined means of passing back pipes / attach / etc?

As for exec, I guess my understanding is you now have the kubelet inject an `exec-shim` into the pod (let's say we run a container that contains `/exec-shim` only), and now the workflow for a user exec operation is the following:

1) User calls `kubectl exec`, a custom already-defined streaming protocol is opened from kubectl to the api-server
2) api-server calls kubelet exec endpoint and proxies reads/writes, leaving the protocol unchanged
3) The kubelet uses a container-runtime provided means to call `exec-shim exec <container-mount-namespace> <command>`
4) `exec-shim` essentially forks off a `nsenter -m=<namespace> -- /command`, creating a pty for it that it controls
5) kubelet pipes the api-server provided stream unmodified into `exec-shim`'s stdin/stdout
6) `exec-shim` understands the api-server protocol and handles all terminal resizes and whatever.

One thing to point out is step 3. The kubelet uses a container-runtime provided mean to call the binary (so now you have an exec call) which we have to specify, and that must be specified independent of other exec path since it is needed to implement it. Furthermore, it has to be executed "remotely" to support the client/server design of CRI this issue deals with. In the next step, we again have to pass the `exec-shim` input/output via CRI, so we have to solve the problem of sending a byte stream anyways, though that still might be worth it.

Am I missing something with the steps above? Is that an accurate assessment?


@mrunalp 
> That implies that we want to follow the docker model and use pipes. The API could then return the path to the pipes as @yujuhong suggested earlier.

Returning paths to pipes doesn't work for hypervisor based runtimes though, right? Also, paths for those sorts of features make it harder to containerize the kubelet.

**Comment by yifan-gu** [comment]
So for exec/portforward, they are easier, we can use nsenter, socat, so it can be runtime agnostic.

Attach is problematic, we either need a parent process who holds the pty and process the data, or doing something really [**interesting**](https://blog.nelhage.com/2011/02/changing-ctty/)...

ref https://github.com/systemd/systemd/issues/2069#issuecomment-160693254

**Comment by yujuhong** [comment]
> So for exec/portforward, they are easier, we can use nsenter, socat, so it can be runtime agnostic.

Those don't work with VMs. Hence the more complicated proposal.

**Comment by yifan-gu** [comment]
Sorry, I forget that...

**Comment by yujuhong** [comment]
@euank let's assume that there is a custom, statically linked binary called kubelet-shim, and kubelet can inject kubelet-shim into the container (perhaps via mounts, or via infra container; the latter is not portable though).

My understanding of how this works (correct me if I am wrong):
 1. User calls `kubectl exec`, and connects to the apiserver.
 2. apiserver reaches kubelet through the http exec endpoint.
 3. kubelet calls `Exec` in CRI to run `kubelet-shim`, which will accept connection at port N. This `Exec` call will not need streaming.
 4. kubelet redirects apiserver to talk to `<pod IP>:<port N>` directly. After this, kubelet will not be involved in the connection at all and all the resource usage of the kubelet-shim is charged to the pod.

The same applies to port-forwarding, but attach may need more work. 
I am not sure exactly how to do the handoff. The approach of injecting the binary also warrants more thoughts.

**Comment by mrunalp** [comment]
The shim need not be injected into the container. It could be the process that launches the container and monitors it.

**Comment by yujuhong** [comment]
> The shim need not be injected into the container. It could be the process that launches the container and monitors it.

How does this work with docker?

**Comment by smarterclayton** [comment]
I believe the best way to do handoff is to continue to use our existing
protocol for exec rather than invent a new one - so a redirect from kubelet
to shim port, followed by an upgrade request from the master.

Agree attach is more problematic.  We *could* come up with a way that
attach is possible via regular containers (via shared filesystem and a side
car container), where we (or the user) starts the exec-shim using a
filesystem pipe.  Doesn't help tty I don't think (or does it?)

**Comment by feiskyer** [comment]
> The protocol used by the client and master, or master and kubelet today is
> our SPDY or WS protocol.  We do not want to continue sending attach and
> exec through the kubelet, nor do we want the runtime daemons proxying this
> data in the long term (that means the network and CPU bandwidth defined by
> exec are not accounted to the container).  The kubelet needs to authorize
> the request and then hand it off to the container.  So ultimately, the
> streaming protocol can be completely separated from the transaction of
> establishing the transaction protocol.

@smarterclayton Is this means runtimes should export public apis to kube-apiserver?

> I don't think CRI should define exec/attach/portforward *streaming* - it
> needs to define establishing that streaming, and then let the container
> shim process manage the subsequent connection.

@smarterclayton The streaming is not always established, only when exec/attach/portforward are called.

https://github.com/kubernetes/kubernetes/issues/29579#issuecomment-239052707 @yujuhong Introducing `kubelet-shim` is making assumptions pod network is always reachable from host network, while this is not always true. With network policy enabled, a namespace could have its own isolated network.

I still think adding streaming apis in CRI (https://github.com/kubernetes/kubernetes/issues/29579#issuecomment-238763699) is the simplest way.

**Comment by smarterclayton** [comment]
I disagree - that makes CRI the bottleneck for potentially unbounded
network traffic, which really doesn't fit with our long term goals for the
node.  I think there could be an exception for isolated networks, but
that's still solvable by having the exec shim not be fully in the
containers network namespace.

**Comment by smarterclayton** [comment]
And to be clear, I don't think all runtimes have to have a shim.  But I
don't see any reason for containers not to have one.

**Comment by euank** [comment]
Injecting a binary sounds easy for exec/port-forward; running a new container (the "kubelet-shim" or "kubelet-helper" container is easy, and has clearly defined behavior (right network namespace, able to be privileged and enter into another mount namespace for exec). That should be portable, and as a bonus accounts the resources to the pod cgroup trivially.

> kubelet calls Exec in CRI to run kubelet-shim, which will accept connection at port N. This Exec call will not need streaming.

I don't like including such a process that is listening on the node. It's arbitrary-code-execution-as-an-endpoint and networking is hard. If it listens on localhost then we need port-forward to be solved which loses the purpose of this.

I think it's preferable to have the `Exec` support streaming of only stdin/stdout and implement the api-server  protocol on top of that much simpler bidirectional stream to include additional streams, ioctls, and what have you.

Alternately, we could bindmount a unix socket. Before I said I didn't think that would work with hypervisor runtimes, but it looks like I was totally off the mark and it can work just fine, so that solves much of my concern there.

**Comment by thockin** [comment]
I am trying to follow this but I lack context.  Is there a higher-level
explanation of what the design is and (more importantly) why this is such a
complex picture?

**Comment by yujuhong** [comment]
> I am trying to follow this but I lack context.  Is there a higher-level
> explanation of what the design is and (more importantly) why this is such a
> complex picture?

I think it's complex because we are trying to tackle several issues altogether, and I am not sure if we need to solve all of them at once (but they do affect each other down the line). 

1. Today, kubelet *is* involved in streaming. @smarterclayton made a point that we should *not* let kubelet become the network bottleneck. (also the resource usage of such connections should be charged to the pod properly). This rules out any streaming support in CRI and opens a whole new discussion of how to achieve that.
2. Today kubelet relies on other tools (e.g., `nsenter` and `socat`) for these operations, and those don't work with hypervised runtimes. We either need a more portable solution (e.g., a static binary discussed above), or we need to push this down to runtimes completely and define all of them in CRI
3. Attach has always been problematic for runtimes to implement, and on top of that we have the streaming issue (1).

**Comment by smarterclayton** [comment]
We already today can DoS a Kubelet with exec / logs / portforward.  We
don't *have* to solve that here (for CRI to land) - we could
potentially leave them grandfathered in the Kubelet until we get there
right solution.  That gives us some flexibility to solve the problem
the right way.

I don't think exec or portforward are optional to anyone except very
specialized end users (they are core enablers for debugging and
general use).  I tend to think attach is required for some use cases,
but not all.  If we agree on that, then we should probably figure out
how to do them in the way that solves accounting (people are already
sending 5gb dumps out of containers in production via exec, and I know
of some people sending large dumps in for restore).

**Comment by feiskyer** [comment]
Should network bottleneck problem be included in this issue? I think it shouldn't be a block for CRI and suggest it in another issue.

I don't think adding `shim` solves the bottleneck/DoS problem. Since all exec/logs request must finally get response from container runtime on node, adding `shim` just moves the network bottleneck to `shim` from kubelet.

**Comment by smarterclayton** [comment]
The point of shim would be that exec/logs would *not* go through the
kubelet.  Instead, the kubelet authorizes the request and delegates to the
shim, and then leaves the communication loop.

My recommendation at this point is to strike exec/attach from the CRI API,
continue to use the existing abstractions in the kubelet for Docker and
Rocket, and open a follow up issue that can be CRI v2 to solve
exec/portforward.  All other runtimes except Docker and Rocket would not be
able to exec/portforward (i.e., be extremely user unfriendly), so we'd be
under pressure to fix this, but we should fix it the correct way.

I haven't heard anyone make a good argument why 3GB backups should be going
through the kubelet, so any design proposal ultimately has to acknowledge
the same limitations for the various CRI daemons.

**Comment by yujuhong** [comment]
How do we handle the authentication between the apiserver and the shim process? Maybe we can ask the process to check the source IP of the request to ensure that it matches the apiserver's. @smarterclayton any suggestions?

**Comment by vishh** [comment]
Without an Exec API, how would the kubelet invoke the injected binary
inside of the containers sandbox?

**Comment by smarterclayton** [comment]
The previous discussion before would be that the shim would expose a
standard HTTP2 interface and would be started with a unique one time token
generated by the Kubelet.  The token would be returned to the master via a
response, or via a redirect.  There's a lot of open questions around "exec
sessions" which is why this needs a proper design.

Most CRI implementations can launch a process in a container sandbox as
part of the initialization of a session.

The interface for launching a session on the CRI might be:

type NewExecSessionRequest struct {
  ... process definition
}
type NewExecSessionResponse struct {
  Token string // unique credential
  HostPort string // must be visible to master
}

The kubelet would then redirect.  A naive CRI could start the process on
the host network (bound to an unused host port) but charged to the cgroup
of the pod/container, and the child of the shim process would be fully in
the network namespace.

**Comment by vishh** [comment]
Ok. Essentially you are proposing adding a specialized `Exec` interface
rather than a general purpose one for bootstrapping.
Given the extent of changes required for this feature, would it make sense
to support regular exec in the short-term before we can replace that with
the proxy process approach?

**Comment by smarterclayton** [comment]
Yes, I think short term exec "as it is today" is the current solution.
Which means not trying to add it to CRI, but leave it in the daemon.

**Comment by smarterclayton** [comment]
s/current/correct/

**Comment by vishh** [comment]
Wait, how would it work if the CRI does not have a concept of Exec? Today's
exec model is one where runtimes implement Exec natively. Kubelet needs
some means to invoke that Exec on the runtimes?

**Comment by smarterclayton** [comment]
To move exec to the CRI means implementing a 3rd hop in the chain between
the master and the container, which makes this problem 50% worse.  So
putting exec into CRI in the current form doesn't actually help anyone.
Instead, we have two options.

1. Handle the full, complex problem by letting the Kubelet hand off exec to
a target CRI destination port, securely.
2. Leave exec / attach / port forward as-is in the Kubelet, implement them
only for Rkt and Docker, and come back and solve 1 later.

**Comment by yujuhong** [comment]
I wrote up a summary with options. PTAL.
https://docs.google.com/document/d/1MreuHzNvkBW6q7o_zehm1CBOBof3shbtMTGtUpjpRmY/edit?usp=sharing

/cc @thockin @dchen1107

**Comment by thockin** [comment]
Great doc, thanks.  I concur with the conclusion

**Comment by feiskyer** [comment]
> I wrote up a summary with options. PTAL.
> https://docs.google.com/document/d/1MreuHzNvkBW6q7o_zehm1CBOBof3shbtMTGtUpjpRmY/edit?usp=sharing

Great summary.

**Comment by smarterclayton** [comment]
I also concur with the conclusion.  @kubernetes/sig-auth needs to
review the security implications and @kubernetes/sig-api-machinery
needs to be involved in final design for anything on the apiserver.

**Comment by erictune** [comment]
Just skimmed this issue. What is the one-time token for?  Is it for the apiserver to prove its identity to the serving process, or for the serving process to prove its identity to the master?  

For master to prove identity to the serving process, can we have the master present a client cert with a well-known CN which identifies it as a trusted proxy for the cluster?  See https://github.com/kubernetes/kubernetes/pull/26634

**Comment by smarterclayton** [comment]
I think there is value in requiring the master to access the kubelet first
in order to establish an exec session.  It prevents some abuse (potentially
programmer error) where the master tries to connect to those shims.  All
that being said though, the master -> shim connection must be HTTPS in
practice and so the shim needs a serving cert.

**Comment by erictune** [comment]
Agree with what you said.  I think the shim still needs to be sure that the thing connecting to it is really the master, and I was suggesting client cert as the way to ensure that.

**Comment by euank** [comment]
@smarterclayton 
I'd like to disagree with one aspect of the conclusion we seem to have reached here: whether we should implement a through-kubelet intermediate or not.

Because CRI is a semi-internal interface, we don't have to feel as bad about including an rpc that we know will be deprecated. I think that it makes sense to include an "option 1" solution while we work towards "option 3", so long as we can make that option 1 solution not require any changes other than kubelet and below (which I think we can do).

The primary reason I argue for that is so that it is easier to run the conformance and e2e tests against these runtimes and so that users may kick the tires of the runtimes with fewer blockers.

If we can manage option 3 quick enough, then we can skip an intermediate option 1 solution, but if we'll have the rest of CRI functionality well before we have option 3, I'd rather have an intermediate option 1 than nothing.

**Comment by smarterclayton** [comment]
If we think having an RPC that doesn't do anything (except return a fake
redirect) that's fine.  The exec protocol is very complicated though, so
anything that has to perform a complex wiring and transformation of those
streams is going to be more work than it's worth.  Only something that is
trivially passthrough is going to be worth the intermediate solution.

**Comment by smarterclayton** [comment]
And note I would not run that option 1 in anything resembling a production
environment, and if we even think that is a possibility then that would be
another argument against an option 1.

**Comment by euank** [comment]
@smarterclayton I'd argue that attach shouldn't be supported in a production environment at all.

The status quo today is that all of those options are implemented as piping through kubelet, so if you would argue that we should avoid implementing them at all in that way, perhaps we should just rip that out of kubelet now out of fear someone might accidentally use them in a production environment.

**Comment by yujuhong** [comment]
I think we are going to attempt option (3) for v1.5  (@timstclair and someone from redhat? will work on this), but worst case we'll fall back to option (1). @smarterclayton, I don't think is option (1) is unacceptable, and we also have in-process docker integration to tie us over until we can implement other options.

@thockin, @dchen1107, correct me if I am wrong. We should finalize this for 1.5 planning.

**Comment by thockin** [comment]
That sounds good.  We can simply timebox a proof-of-concept, and if that doesn't seem on-track, fall back on (1) which seems easiest.

**Comment by smarterclayton** [comment]
Has anyone actually looked at 1?  The overhead of that seems high for small messages, plus the addition of a hop.  I'd be a little cautious about saying it's easiest unless we've actually looked at the code on both ends.   Last I saw gRPC wasn't much more efficient than HTTP - has that improved?

**Comment by thockin** [comment]
It's not about efficiency to get to alpha - it's about proving that the
overall model works.  We should not block 97 APIs because 3 of them are
imperfect.

**Comment by smarterclayton** [comment]
Attach is irrelevant - exec and port forward are important tools.  There's
a really big difference between using them in production (which we have
people doing) and ​adding a) a brand new buggy implementation and b) making
it inefficient and c) adding an extra hop to it.  To an alpha release that
is just a test - sure.  But that alpha release isn't something I'd use in a
real environment, so it seems strange to rush to put it into an interface
if the design is wrong for the long term goals.

We can revisit after we find out whether 3 works.

**Comment by feiskyer** [comment]
> That sounds good. We can simply timebox a proof-of-concept, and if that doesn't seem on-track, fall back on (1) which seems easiest.

Agree with @thockin. Need a POC of option(3) first before making a final decision (even for the alpha API).

**Comment by lucab** [comment]
While implementing the attach/exec part for rkt, I had some concerns related to streams/tty/VMs interaction that I summarized here https://github.com/kubernetes/kubernetes/issues/32869. In the context of @timstclair work on this, I'd gladly have a more simpler abstraction which doesn't expose single streams for out and err across host/pod border.

**Comment by ncdc** [comment]
FYI, one other thing we may need to consider is whether or not attach should include or exclude the container's logs. For example, if you run a container that prints some lines to stdout, `docker attach mycontainer` will not show anything printed to stdout prior to attaching. Any new output printed to stdout will show up in the attach session. There is, however, a feature that the docker daemon supports where, when attaching, you can get any output previously logged. We actually used to use this feature when we were still using go-dockerclient, but it went away when we switched to docker's engine-api, because it doesn't support it. I have created https://github.com/docker/docker/pull/26718 to add `Logs bool` to `ContainerAttachOptions` so we can restore this functionality if docker merges the PR and if we decide we want to display logs when attaching.

**Comment by timstclair** [comment]
Design doc: https://docs.google.com/document/d/1OE_QoInPlVCK9rMAx9aybRmgFiVjHpJCHI9LrfdNM_s/edit?usp=sharing

There shouldn't be any surprises there, it's the same design we had previously discussed (option 3), but with more details filled in.

**Comment by euank** [comment]
Well, we've started implementing from that design-doc. I think we can close this issue and carry on in others.

**Comment by timstclair** [comment]
I'm still using this as the umbrella issue. I changed the title to not be a question.

**Comment by euank** [comment]
If this is meant to be an umbrella issue we should link issues to it / link to issues from it.

Neither of those has happened for the many recent PRs related to this implementation, so it felt more like a :closed_umbrella: than :open_umbrella:

**Comment by timstclair** [comment]
Yeah, I should have been linking PRs to an issue. I just back-filled. :open_umbrella: :+1: :question:

**Comment by timstclair** [comment]
I think we can close this now, since the indirect (CRI) path is setup for the dockershim. Remaining work includes:

- Resolving TTY in Kubelet (https://github.com/kubernetes/kubernetes/pull/36615)
- Deal with very long commands (https://github.com/kubernetes/kubernetes/issues/36187)
- Implement authentication interface (https://github.com/kubernetes/kubernetes/issues/36666)
- Serve streaming commands from Pod namespace (https://github.com/kubernetes/kubernetes/issues/36667)
